spec/jobs/tweet_fetching_job_spec.rb

```diff
@@ -0,0 +1,179 @@
+require 'rails_helper'
+
+describe TweetFetchingJob do
+  let(:ds) { build(:data_set) }
+
+  # rspec overall needs to execute jobs immediately so that data-collecting
+  # behavior happens, but for the duration of this test suite we need to use
+  # the default job execution behavior, so that we can test it.
+  before(:all) { Delayed::Worker.delay_jobs = true }
+  after(:all)  { Delayed::Worker.delay_jobs = false }
+
+  context 'during data ingestion', elasticsearch: true do
+    it 'asks twitter for data on a user' do
+      VCR.use_cassette('data set spec') do
+        expect_any_instance_of(Twitter::REST::Client)
+          .to receive(:user_timeline)
+          .once
+          .with(1, count: Rails.application.config.tweets_per_user,
+                tweet_mode: 'extended')
+
+        allow_any_instance_of(TweetFetchingJob).to receive(:store_data)
+        allow(ds).to receive(:finalize_when_ready)
+
+        TweetFetchingJob.perform_now(ds, 1)
+      end
+    end
+
+    it 'creates an elasticsearch document for each tweet' do
+      tweets = [
+        { foo: 1 }, { bar: 2 }
+      ]
+      allow_any_instance_of(TweetFetchingJob).to receive(:fetch_tweets)
+                                             .and_return(tweets)
+      verify_index
+
+      # Why not expect_any_instance_of(Elasticsearch::Client)? Because the
+      # create action is provided by a mixin, so although it's present on all
+      # instances, the the test suite can't find it on the class, and therefore
+      # can't mock it.
+      expect_any_instance_of(Elasticsearch::API::Actions)
+        .to receive(:create)
+        .once.with(index: ds.index_name, type: '_doc', body: tweets[0].to_json)
+      expect_any_instance_of(Elasticsearch::API::Actions)
+        .to receive(:create)
+        .once.with(index: ds.index_name, type: '_doc', body: tweets[1].to_json)
+
+      TweetFetchingJob.perform_now(ds, 1)
+    end
+
+    it 'handles unauthorized accounts' do
+      allow_any_instance_of(TweetFetchingJob).to receive(:fetch_tweets)
+        .and_raise Twitter::Error::Unauthorized
+      verify_index
+
+      TweetFetchingJob.perform_now(ds, 1)
+      TweetFetchingJob.perform_now(ds, 2)
+
+      expect(ds.unauthorized).to match_array ['1', '2']
+    end
+  end
+
+  context 'backoff' do
+    it 'is none when is nothing else in the queue' do
+      mock_count(1)
+
+      expect(TweetFetchingJob.backoff).to eq 0
+    end
+
+    it 'is none when there is not much in the queue' do
+      mock_count(Rails.configuration.rate_limit_limit * 0.25)
```

Comment:
It feels weird to me to have the outcome of simple, unit-like tests depend on app config values rather than be fully deterministic.

Reply:
These need to be app configuration values because the the test semantics aren't "have we achieved a specific number" but "have we abided by Twitter's policy". Twitter could change that policy at any time. Having it in config makes it easy for us to change anywhere it appears in the codebase by ensuring there's only one place we need to change.

Reply:
👍